#!/bin/bash

if [ $# -ne 3 ]; then
    echo "usage: $0 sample_name input_vcf output_file"
    echo "input_vcf must include a sample named 'phasedgt', which is"
    echo "usually generated by first running SHAPEIT2 on bulk and"
    echo "then joining the phased VCF with a single cell VCF."
    exit 1
fi

sample_name=$1
input_vcf=$2
output_file=$3

# the only called genotypes in phasedgt are phased and heterozygous
java -jar ~/balance/resources/GATK3.8.jar -T SelectVariants \
    -R ~/balance/resources/human_g1k_v37_decoy.fasta \
    -sn $sample_name \
    -sn phasedgt \
    -V $input_vcf \
    -env -trimAlternates \
    -select 'vc.getGenotype("'$sample_name'").isCalled()' \
    -select 'vc.getGenotype("phasedgt").isCalled()' \
    -select 'vc.isBiallelic()' \
    -selectType SNP \
    -o $output_file
